File: onyx-database-tests/test/remote/SocketSerializerTest.java
package remote;

import category.RemoteServerTests;
import com.onyx.map.serializer.SocketBuffer;
import gnu.trove.THashMap;
import org.junit.Assert;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import pojo.*;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;

/**
 * Created by timothy.osborn on 4/14/15.
 */
@Category({ RemoteServerTests.class })
public class SocketSerializerTest
{
    @Test
    public void testSimple() throws IOException
    {
        Simple instance = new Simple();
        instance.hiya = 4;

        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        Simple instance2 = (Simple)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance.hiya == instance2.hiya);
    }

    @Test
    public void testAllValues() throws IOException
    {
        AllTypes instance = new AllTypes();
        instance.intValue = 654;
        instance.intValueM = 246;
        instance.longValue = 998l;
        instance.longValueM = 999l ;
        instance.booleanValue = false;
        instance.booleanValueM = true;
        instance.shortValue = 44;
        instance.shortValueM = 45;
        instance.doubleValue = 12.123d;
        instance.doubleValueM = 23.124d;
        instance.floatValue = 11.2f;
        instance.floatValueM = 23.2f;
        instance.byteValue = ((byte)2);
        instance.byteValueM = Byte.valueOf((byte)3);
        instance.dateValue = new Date(3988);
        instance.stringValue = "Test String";
        instance.nullValue = null;
        instance.charValue = 'A';

        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        AllTypes instance2 = (AllTypes)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance.intValue == instance2.intValue);
        Assert.assertTrue(instance.intValueM.equals(instance2.intValueM));
        Assert.assertTrue(instance.longValue == instance2.longValue);
        Assert.assertTrue(instance.longValueM.equals(instance2.longValueM));
        Assert.assertTrue(instance.booleanValue == instance2.booleanValue);
        Assert.assertTrue(instance.booleanValueM.equals(instance2.booleanValueM));
        Assert.assertTrue(instance.shortValueM.equals(instance2.shortValueM));
        Assert.assertTrue(instance.doubleValue == instance2.doubleValue);
        Assert.assertTrue(instance.shortValue == instance2.shortValue);
        Assert.assertTrue(instance.doubleValueM.equals(instance2.doubleValueM));
        Assert.assertTrue(instance.floatValue == instance2.floatValue);
        Assert.assertTrue(instance.floatValueM.equals(instance2.floatValueM));
        Assert.assertTrue(instance.byteValue == instance2.byteValue);
        Assert.assertTrue(instance.byteValueM.equals(instance2.byteValueM));
        Assert.assertTrue(instance.dateValue.getTime() == instance2.dateValue.getTime());
        Assert.assertTrue(instance.stringValue.equals(instance2.stringValue));
        Assert.assertTrue(instance.nullValue == instance2.nullValue);
        Assert.assertTrue(instance.charValue == instance2.charValue);
    }

    @Test
    public void testEnum() throws IOException
    {
        EnumTypeObject instance = new EnumTypeObject();
        instance.intValue = 44;
        instance.simpleEnum = SimpleEnum.SECOND;
        instance.longValue = 234235245l;

        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        EnumTypeObject instance2 = (EnumTypeObject)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance.intValue == instance2.intValue);
        Assert.assertTrue(instance.longValue == instance2.longValue);
        Assert.assertTrue(instance.simpleEnum == instance2.simpleEnum);
    }

    @Test
    public void testTransient() throws IOException
    {
        TransientValue instance = new TransientValue();
        instance.intValue = 44;
        instance.longValue = 234l;
        instance.zdateValue = new Date(23423);

        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        TransientValue instance2 = (TransientValue)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance.intValue == instance2.intValue);
        Assert.assertTrue(instance.longValue != instance2.longValue);
        Assert.assertTrue(instance.zdateValue.getTime() == instance2.zdateValue.getTime());
    }

    
    @Test
    public void testArrayObject() throws IOException
    {
        ArrayObject instance = new ArrayObject();
        instance.longArray = new Long[3];
        instance.objectArray = new Object[4];
        instance.simpleArray = new Simple[2];

        instance.longArray[0] = 223l;
        instance.longArray[1] = 293l;
        instance.longArray[2] = 323l;

        AllTypes obj = new AllTypes();
        obj.intValue = 23;

        instance.objectArray[3] = obj;

        instance.simpleArray[1] = new Simple();
        instance.simpleArray[1].hiya = 99;


        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        ArrayObject instance2 = (ArrayObject)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance2.longArray[0] == 223l);
        Assert.assertTrue(instance2.longArray[1] == 293l);
        Assert.assertTrue(instance2.longArray[2] == 323l);

        Assert.assertTrue(instance2.objectArray[0] == null);
        Assert.assertTrue(instance2.objectArray[1] == null);
        Assert.assertTrue(instance2.objectArray[3] instanceof AllTypes);
        Assert.assertTrue(((AllTypes)instance2.objectArray[3]).intValue == 23);

        Assert.assertTrue(instance2.simpleArray[0] == null);
        Assert.assertTrue(instance2.simpleArray[1] instanceof Simple);
        Assert.assertTrue(instance2.simpleArray[1].hiya == 99);

    }

    @Test
    public void testListObject() throws IOException
    {
        ListObject instance = new ListObject();
        instance.longArray = new ArrayList();
        instance.objectArray = new ArrayList();
        instance.simpleArray = new ArrayList();

        instance.longArray.add(223l);
        instance.longArray.add(293l);
        instance.longArray.add(323l);

        AllTypes obj = new AllTypes();
        obj.intValue = 23;

        instance.objectArray.add(obj);

        instance.simpleArray.add(new Simple());
        instance.simpleArray.get(0).hiya = 99;


        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        ListObject instance2 = (ListObject)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance2.longArray.get(0) == 223l);
        Assert.assertTrue(instance2.longArray.get(1) == 293l);
        Assert.assertTrue(instance2.longArray.get(2) == 323l);

        Assert.assertTrue(instance2.objectArray.get(0) instanceof AllTypes);
        Assert.assertTrue(((AllTypes)instance2.objectArray.get(0)).intValue == 23);

        Assert.assertTrue(instance2.simpleArray.get(0) instanceof Simple);
        Assert.assertTrue(instance2.simpleArray.get(0).hiya == 99);

    }

    @Test
    public void testMapSerialization() throws IOException
    {
        MapObject instance = new MapObject();
        instance.simpleMap = new HashMap();

        instance.simpleMap.put("NEW", new Simple());
        instance.simpleMap.put("NEW2", new Simple());
        instance.simpleMap.put("NEW3", new Simple());
        instance.simpleMap.put("NEW4", null);

        instance.simpleMap.get("NEW").hiya = 2324;
        instance.simpleMap.get("NEW3").hiya = 2924;

        instance.objectMap = new THashMap();
        instance.objectMap.put(new Long(23), new Integer(22));
        instance.objectMap.put(new Integer(12), false);
        instance.objectMap.put(new Simple(), new AllTypes());

        final ByteBuffer buffer = SocketBuffer.serialize(instance);
        buffer.rewind();

        MapObject instance2 = (MapObject)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance2.simpleMap.size() == 4);
        Assert.assertTrue(instance2.simpleMap.get("NEW").hiya == 2324);
        Assert.assertTrue(instance2.simpleMap.get("NEW3").hiya == 2924);
        Assert.assertTrue(instance2.simpleMap.get("NEW2").hiya == 3);
        Assert.assertTrue(instance2.simpleMap.get("NEW4") == null);

        Assert.assertTrue(instance2.objectMap.size() == 3);
        Assert.assertTrue(instance2.objectMap.get(new Long(23)).equals(22));
        Assert.assertTrue(instance2.objectMap.get(new Simple()).equals(new AllTypes()));
        Assert.assertTrue((boolean)instance2.objectMap.get(new Integer(12)) == false);
    }

    @Test
    public void testComplexObject() throws IOException
    {
        ComplexObject object = new ComplexObject();
        object.child = new ComplexObjectChild();
        object.child.parent = object;
        object.mine = object;

        object.dateValue = new Date(23423);
        object.child.longValue = 33l;

        final ByteBuffer buffer = SocketBuffer.serialize(object);
        buffer.rewind();

        ComplexObject instance2 = (ComplexObject)SocketBuffer.deserialize(buffer);

        Assert.assertTrue(instance2.dateValue.getTime() == object.dateValue.getTime());
        Assert.assertTrue(instance2.child.longValue == object.child.longValue);
        Assert.assertTrue(instance2.child.parent == instance2);
        Assert.assertTrue(instance2.mine == instance2);
    }

}

File: onyx-database-tests/test/serialization/SocketSerializerTest.kt
package serialization

import com.onyx.buffer.BufferStream
import org.junit.Test
import pojo.*

import java.util.ArrayList
import java.util.Date
import java.util.HashMap
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Created by timothy.osborn on 4/14/15.
 */
class SocketSerializerTest {

    @Test
    fun testSimple() {
        val instance = Simple()
        instance.hiya = 4

        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as Simple?
        assertEquals(instance.hiya, instance2!!.hiya)
    }

    @Test
    fun testAllValues() {
        val instance = AllTypes()
        instance.intValue = 654
        instance.intValueM = 246
        instance.longValue = 998L
        instance.longValueM = 999L
        instance.booleanValue = false
        instance.booleanValueM = true
        instance.shortValue = 44
        instance.shortValueM = 45
        instance.doubleValue = 12.123
        instance.doubleValueM = 23.124
        instance.floatValue = 11.2f
        instance.floatValueM = 23.2f
        instance.byteValue = 2.toByte()
        instance.byteValueM = java.lang.Byte.valueOf(3.toByte())
        instance.dateValue = Date(3988)
        instance.stringValue = "Test String"
        instance.nullValue = null
        instance.charValue = 'A'

        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as AllTypes?

        assertEquals(instance.intValue, instance2!!.intValue)
        assertEquals(instance.intValueM, instance2.intValueM)
        assertEquals(instance.longValue, instance2.longValue)
        assertEquals(instance.longValueM, instance2.longValueM)
        assertEquals(instance.booleanValue, instance2.booleanValue)
        assertEquals(instance.booleanValueM, instance2.booleanValueM)
        assertEquals(instance.shortValueM, instance2.shortValueM)
        assertEquals(instance.doubleValue, instance2.doubleValue)
        assertEquals(instance.shortValue, instance2.shortValue)
        assertEquals(instance.doubleValueM, instance2.doubleValueM)
        assertEquals(instance.floatValue, instance2.floatValue)
        assertEquals(instance.floatValueM, instance2.floatValueM)
        assertEquals(instance.byteValue, instance2.byteValue)
        assertEquals(instance.byteValueM, instance2.byteValueM)
        assertEquals(instance.dateValue!!.time, instance2.dateValue!!.time)
        assertEquals(instance.stringValue, instance2.stringValue)
        assertEquals(instance.nullValue, instance2.nullValue)
        assertEquals(instance.charValue, instance2.charValue)
    }

    @Test
    fun testEnum() {
        val instance = EnumTypeObject()
        instance.intValue = 44
        instance.simpleEnum = SimpleEnum.SECOND
        instance.longValue = 234235245L

        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as EnumTypeObject?

        assertEquals(instance.intValue, instance2!!.intValue)
        assertEquals(instance.longValue, instance2.longValue)
        assertEquals(instance.simpleEnum, instance2.simpleEnum)
    }

    @Test
    fun testTransient() {
        val instance = TransientValue()
        instance.intValue = 44
        instance.longValue = 234L
        instance.zDateValue = Date(23423)

        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as TransientValue?

        assertEquals(instance.intValue, instance2!!.intValue)
        assertNotEquals(instance.longValue, instance2.longValue)
        assertEquals(instance.zDateValue!!.time, instance2.zDateValue!!.time)
    }


    @Test
    fun testArrayObject() {
        val instance = ArrayObject()
        instance.longArray = arrayOfNulls(3)
        instance.objectArray = arrayOfNulls(4)
        instance.simpleArray = arrayOfNulls(2)

        instance.longArray!![0] = 223L
        instance.longArray!![1] = 293L
        instance.longArray!![2] = 323L

        val obj = AllTypes()
        obj.intValue = 23

        instance.objectArray!![3] = obj

        instance.simpleArray!![1] = Simple()
        instance.simpleArray!![1]!!.hiya = 99


        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as ArrayObject?

        assertEquals(223L, instance2!!.longArray!![0])
        assertEquals(293L, instance2.longArray!![1])
        assertEquals(323L, instance2.longArray!![2])

        assertNull(instance2.objectArray!![0])
        assertNull(instance2.objectArray!![1])
        assertTrue(instance2.objectArray!![3] is AllTypes)
        assertEquals(23, (instance2.objectArray!![3] as AllTypes).intValue)

        assertNull(instance2.simpleArray!![0])
        assertTrue(instance2.simpleArray!![1] is Simple)
        assertEquals(99, instance2.simpleArray!![1]!!.hiya)
    }

    @Test
    fun testListObject() {
        val instance = ListObject()
        instance.longArray = ArrayList()
        instance.objectArray = ArrayList()
        instance.simpleArray = ArrayList()

        instance.longArray!!.add(223L)
        instance.longArray!!.add(293L)
        instance.longArray!!.add(323L)

        val obj = AllTypes()
        obj.intValue = 23

        instance.objectArray!!.add(obj)

        instance.simpleArray!!.add(Simple())
        instance.simpleArray!![0].hiya = 99


        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as ListObject?

        assertEquals(223L, instance2!!.longArray!![0])
        assertEquals(293L, instance2.longArray!![1])
        assertEquals(323L, instance2.longArray!![2])
        assertTrue(instance2.objectArray!![0] is AllTypes)
        assertEquals(23, (instance2.objectArray!![0] as AllTypes).intValue)
        assertEquals(99, instance2.simpleArray!![0].hiya)

    }

    @Test
    fun testMapSerialization() {
        val instance = MapObject()
        instance.simpleMap = HashMap()

        instance.simpleMap!!.put("NEW", Simple())
        instance.simpleMap!!.put("NEW2", Simple())
        instance.simpleMap!!.put("NEW3", Simple())
        instance.simpleMap!!.put("NEW4", null)

        instance.simpleMap!!["NEW"]!!.hiya = 2324
        instance.simpleMap!!["NEW3"]!!.hiya = 2924

        instance.objectMap = HashMap()
        instance.objectMap!!.put(23, 22)
        instance.objectMap!!.put(12, false)
        instance.objectMap!!.put(Simple(), AllTypes())

        val buffer = BufferStream.toBuffer(instance)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as MapObject?

        assertEquals(4, instance2!!.simpleMap!!.size)
        assertEquals(2324, instance2.simpleMap!!["NEW"]!!.hiya)
        assertEquals(2924, instance2.simpleMap!!["NEW3"]!!.hiya)
        assertEquals(3, instance2.simpleMap!!["NEW2"]!!.hiya)
        assertNull(instance2.simpleMap!!["NEW4"])

        assertEquals(3, instance2.objectMap!!.size)
        assertEquals(22, instance2.objectMap!![23])
        assertEquals(AllTypes(), instance2.objectMap!![Simple()])
        assertEquals(false, instance2.objectMap!![12])
    }

    @Test
    fun testComplexObject() {
        val `object` = ComplexObject()
        `object`.child = ComplexObjectChild()
        `object`.child!!.parent = `object`
        `object`.mine = `object`

        `object`.dateValue = Date(23423)
        `object`.child!!.longValue = 33L

        val buffer = BufferStream.toBuffer(`object`)
        buffer.rewind()

        val instance2 = BufferStream.fromBuffer(buffer) as ComplexObject?

        assertEquals(instance2!!.dateValue!!.time, `object`.dateValue!!.time)
        assertEquals(instance2.child!!.longValue, `object`.child!!.longValue)
        assertEquals(instance2.child!!.parent, instance2)
        assertEquals(instance2.mine, instance2)
    }

}

